plugins/plugin-core-impl/src/connection.ts

```diff
@@ -12,56 +12,49 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 //
-import { generateId, Storage, Tx } from '@anticrm/core'
-import { readResponse, RequestProcessor, serialize } from '@anticrm/rpc'
+import { Class, Doc, DocumentQuery, Ref, Storage, Tx } from '@anticrm/core'
+import { readResponse, Request, RequestProcessor, serialize } from '@anticrm/rpc'
 
-export async function connect (
-  clientUrl: string,
-  handler: (tx: Tx) => void
-): Promise<Storage> {
-  // Client
-  const clientId = generateId() // <-- Uniq ID to identify client in case of re-connections.
-
-  let socket: Promise<WebSocket>
-
-  function isClosed (socket: WebSocket): boolean {
-    return socket.readyState === WebSocket.CLOSED || socket.readyState === WebSocket.CLOSING
-  }
+export async function connect (clientUrl: string, handler: (tx: Tx) => void): Promise<Storage> {
+  const socket = new WebSocket(`ws://${clientUrl}`)
+  await new Promise<any>((resolve, reject) => {
+    socket.onopen = resolve
+    socket.onerror = (reason) => {
+      reject(new Error(`Failed to connect to ${clientUrl}`))
+    }
+  })
 
-  async function doConnect (): Promise<WebSocket> {
-    try {
-      if (socket === undefined || isClosed(await socket)) {
-        socket = new Promise<WebSocket>(resolve => {
-          const ws = new WebSocket(`ws://${clientUrl}/${clientId}`)
-          ws.onopen = () => {
-            resolve(ws)
+  class WebSocketConnection extends RequestProcessor implements Storage {
+    constructor () {
+      super()
+      socket.onerror = (reason) => this.reject('Connection error')
+      socket.onclose = (reason) => this.reject('Connection closed')
+      socket.onmessage = (evt) => this.onmessage(evt)
+    }
```

Comment:
I think making `socket` and `handler` private fields and moving class declaration out of `connect` would be better to understand it's logic.

Reply:
Moved, Not sure if it make it any more clear :-)